src/type-exercises.ts

```diff
@@ -0,0 +1,262 @@
+ *
+ * @example
+ * type A = If<true, 'a', 'b'>;  // expected to be 'a'
+ * type B = If<false, 'a', 'b'>; // expected to be 'b'
+ */
+
+// Add here your solution
+type If<C extends boolean, T, F> = C extends true ? T : F;
+
+// Add here your example
+console.log('Exercise #2: Implement the utility type `If<C, T, F>` 👀');
+console.log('C is true');
+type A = If<true, 'a', 1>;
+console.log('Type A:', 'a' as A);
+
+console.log('C is false');
+type B = If<false, 'a', 1>;
+console.log('Type B:', 2 as B);
+
+/**
+ * Exercise #3: Recreate the built-in `Readonly<T>` utility type without using it.
+ *
+ * @description
+ * Constructs a type that makes all properties of `T` readonly.
+ * This means the properties of the resulting type cannot be reassigned.
+ *
+ * @example
+ * interface Todo {
+ *   title: string;
+ *   description: string;
+ * }
+ *
+ * const todo: MyReadonly<Todo> = {
+ *   title: "Hey",
+ *   description: "foobar"
+ * };
+ *
+ * todo.title = "Hello";       // Error: cannot reassign a readonly property
+ * todo.description = "barFoo"; // Error: cannot reassign a readonly property
+ */
+
+
+// Add here your solution
+type MyReadonly<T> = {
+    readonly [K in keyof T]: T[K];
+};
+
+interface Todo {
+    title: string;
+    description: string;
+}
+
+const todo: MyReadonly<Todo> = {
+    title: "Hey",
+    description: "foobar"
+};
+
+// Add here your example
+console.log('Exercise #3: Recreate the built-in `Readonly<T>` utility type without using it ☃️');
+// todo.title = 'Hello'; 
+console.log(todo.title);
+
+
+/**
+ * Exercise #4: Recreate the built-in `ReturnType<T>` utility type without using it.
+ *
+ * @description
+ * The `MyReturnType<T>` utility type extracts the return type of a function type `T`.
+ *
+ * @example
+ * const fn = (v: boolean) => {
+ *   if (v) {
+ *     return 1;
+ *   } else {
+ *     return 2;
+ *   }
+ * };
+ *
+ * type a = MyReturnType<typeof fn>; // expected to be "1 | 2"
+ */
+
+// Add here your solution
+// type If<C extends boolean, T, F> = C extends true ? T : F;
+type MyReturnType<T> = T extends (...args: any[]) => infer R ? R : any;
+
+const fn = (v: boolean) => {
+    if (v) {
+        return true;
+    } else {
+        return true;
+    }
+};
+
+// Add here your example
+console.log('Exercise #4: Recreate the built-in `ReturnType<T>` utility type without using it 🚀');
+type custom = MyReturnType<typeof fn>;
+console.log('Type a:', typeof (true as custom));
+
+
+
+/**
+ * Exercise #5: Extract the type inside a wrapped type like `Promise`.
+ *
+ * @description
+ * Implement a utility type `MyAwaited<T>` that retrieves the type wrapped in a `Promise` or similar structure.
+ *
+ * If `T` is `Promise<ExampleType>`, the resulting type should be `ExampleType`.
+ *
+ * @example
+ * type ExampleType = Promise<string>;
+ *
+ * type Result = MyAwaited<ExampleType>; // expected to be "string"
+ */
+
+// Add here your solution
+type MyAwaited<T> = T extends Promise<infer R> ? R : T;
```

Comment:
What would happen if the promise returns another promise? Look into recursive types to solve this issue

Reply:
If that happens, the promise will resolve to another promise, creating a nested promise. To handle this situation effectively and ensure proper resolution, we can use recursive types in TypeScript such as:
```ts
type MyAwaited<TPromise> = TPromise extends Promise<infer TResolved>
  ? MyAwaited<TResolved>
  : TPromise
  ```